_posts/2017-09-21-scalamacros.md

```diff
@@ -0,0 +1,438 @@
+and for which use-cases they're either overkill, or insufficient.
+Macros should in general be used as a last resort.
+We care deeply to know what you'd like to see covered in this effort to map
+available metaprogramming abstractions in Scala.
+Please share your thought.
+
+### Share your feedback
+The current set of approved features in macros v3 does not reach feature parity
+with the scala.reflect-based macro system.
+Some scala.reflect macros rely on advanced capabilities beyond what
+blackbox macros support.
+Most notably, these features include (but are not limited to) whitebox def
+macros and macro annotations.
+
+By not supporting these advanced features, we put ourselves a fragile situation
+where we risk forcing Scala users to remain on old versions of the compiler.
+Such a situation is undesirable for both Scala users and those who wish to
+evolve the language.
+We must debate together as a community whether these advanced features
+merit inclusion in the language specification or if they can be replaced
+with alternative metaprogramming techniques.
+
+Here below I try to summarize the pros and cons of whitebox def macros
+and macro annotations.
+I invite you to invite you to start a thread in [Scala Contributors] and
+share your personal thoughts.
+In particular, we want to explore
+
+- what does your macro do and why is it important for your users?
+- can you use alternative metaprogramming techniques such as
+  code generation scripts or compiler plugins to achieve the same
+  functionality?
+
+#### Whitebox def macros
+
+Whitebox def macros have neither been approved nor rejected for inclusion into
+macros v3.
+Whitebox macros are similar to blackbox def macros with the distinction
+that the result type of whitebox def macros can be refined at each call-site.
+The ability to refine the result types opens up exciting applications including
+- [fundep materialization], used by shapeless Generic
+- [extractor macros], used by quasiquotes in scala.reflect, scala.meta, and
+scala.macros
+- [anonymous type providers]
+
+To give an example of how blackbox and whitebox macros differ, imagine that we
+wish to implement a macro to convert case classes into tuples.
+
+```scala
+import scala.macros._
+object CaseClass {
+  def toTuple[T](e: T): Product = macro { ??? }
+  case class User(name: String, age: Int)
+  // if blackbox: expected (String, Int), got Product
+  // if whitebox: OK
+  val user: (String, Int) = CaseClass.toTuple(User("Jane", 30))
+}
+```
+
+As you can see from this example, whitebox macros are more powerful than
+blackbox def macros.
+A whitebox macro that declares its result type as `Any` can have it's result
+type refined to any precise type in the Scala typing lattice.
+This powerful capability opens up questions.
+For example, do implicit whitebox def macros always need to be expanded in
+order be disqualified as a candidate during implicit search.
```

Comment:
I would put a question mark at the end of this sentence since it is interrogative.